### PR TITLE
Added Burkhard Keller Tree Implementation in C++

### PR DESCRIPTION
**Pull request by HarshCasper**
Fixes [#1767](https://github.com/jainaman224/Algo_Ds_Notes/issues/1767)

#### Checklist:
-  [X] 4 space indentation.
- [X] Coding conventions are followed.
- [ ] Input is taken dynamically.
- [X] Sample Input / Output is added at the end of file.
- [X] Logic Documentation (Comments).
- [X] File names are correct.

#### Changes proposed in this pull request:
- Implementation of Burkhard-Keller Tree.

#### Languages Used:
- C++

#### Files Added:
- bktree.cpp

Thanks!

### CODE REVIEW

**Comment by jainaman224** [comment]
Fix file and folder name.

File: BK_Tree/bktree.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+Burkhard Keller Tree is a popular String Matching Algorithm which is used to
+perform spell checks based on the concept of Levenshtein distance.
+
+*/
+
+#include <iostream>
+#include <vector>
+#include <map>
+
+
```
Comment (SakshiShreya) [inline BK_Tree/bktree.cpp]: Remove extra new lines.

File: BK_Tree/bktree.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+typedef std::vector< std::string > strlist;
+
+// Defining the Node 
+
+class Node {
+    public:
+        std::map< int, Node* > Edge;
```
Comment (SakshiShreya) [inline BK_Tree/bktree.cpp]: Variable names should be in camelCase.

File: BK_Tree/bktree.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    return v0.back();
+}
+
+// Driver Function
+
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    strlist dictionary;
+    dictionary={"hell","help","shel","smell","fell","felt","oops","pop","oouch","halt"};
```
Comment (SakshiShreya) [inline BK_Tree/bktree.cpp]: Add space after comma

**Comment by HarshCasper** [comment]
I have updated the required changes. Can you please look them over @SakshiShreya @jainaman224

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    return v0.back();
+}
+
+// Driver Function
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    strlist dictionary;
+    dictionary={"hell", "help", "shel", "smell", "fell", "felt", "oops", "pop", "oouch", "halt"};
```
Comment (jainaman224) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: Take input from user.

**Comment by somya-kapoor** [comment]
@HarshCasper Are you working on it?

**Comment by HarshCasper** [comment]
I am currently working on it @somya-kapoor. Due to my College Exams, my work pace is low so kindly bear with me for some time.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
+/* 
+
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: This new-line was not-required.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+Burkhard Keller Tree is a popular String Matching Algorithm which is used to
+perform spell checks based on the concept of Levenshtein distance.
+
+*/
+#include <iostream>
+#include <vector>
+#include <map>
+typedef std::vector< std::string > strlist;
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: There should be a new-line after include statements.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+// Defining the Node 
+class Node {
+    public:
+        std::map< int, Node* > TreeEdge;
+        std::string Name;
+};
+// Defining the BK Tree
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: There should be 2 new-lines between 2 classes.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    private:
+        Node* root;
+        void insert(std::string& Name, Node*& node);
+        void search(Node* node, std::string& Name, int target_dist, strlist& suggestion_points);
+        int min(int& a, int& b, int& c);
+        int d(std::string& test, std::string& comp);
+        void delete_node(Node* node);
+};
+
+// Constructor to initialize the BK Tree
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: This comment is for the constructor. Please remove the new line between this comment and the constructor.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+BKTree::BKTree(std::string& Root_STR){
+    root = new Node();
+    root->Name = Root_STR;
+}
+
+// Destructor to remove any memory in the BK Tree
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: Same here. Remove the line between the comment and destructor.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        void insert(std::string& Name, Node*& node);
+        void search(Node* node, std::string& Name, int target_dist, strlist& suggestion_points);
+        int min(int& a, int& b, int& c);
+        int d(std::string& test, std::string& comp);
+        void delete_node(Node* node);
+};
+
+// Constructor to initialize the BK Tree
+
+BKTree::BKTree(std::string& Root_STR){
```
Comment (SakshiShreya) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp]: Add a space between `)` and `{`.

**Comment by somya-kapoor** [comment]
@HarshCasper Are you working on it?

**Comment by HarshCasper** [comment]
I will add all the changes by Sunday, @somya-kapoor

**Comment by HarshCasper** [comment]
I have added the changes. Kindly review it @somya-kapoor

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    dictionary={"hell", "help", "shel", "smell", "fell", "felt", "oops", "pop", "oouch", "halt"};
+
+    tree.insert_into(dictionary);
+    
+    strlist test = tree.search("oo", 2);
+    for(auto& i : test)
+        std::cout << i << std::endl;
+    return 0;
+}
+/*
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:156]: Leave a line before it.

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp (hunk 4 above)
Comment (somya-kapoor) [reply Burkhard_Keller_Tree/burkhard_keller_tree.cpp:147]: Solve this issue

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        node->Name = Name;
+    }
+    else {
+        int dist = d(Name, node->Name);
+        insert(Name, node->TreeEdge[dist]);
+    }
+}
+
+// Deletion Operation in BK Tree
+
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:68]: remove this line

**Comment by HarshCasper** [comment]
I have added the changes. Kindly check it out @somya-kapoor @SakshiShreya

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    std::string input;
+    while(std::cin>>input){
+      if(input=="1") {break;}
+      dictionary.push_back(input);
+    }
+
+    tree.insert_into(dictionary);
+    std::string searchStr;
+    std::cout<<"Enter the String that needs to be checked:"<<std::endl;
+    std::cin>>searchStr;
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:153]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+    dictionary={};
+    std::string input;
+    while(std::cin>>input){
+      if(input=="1") {break;}
+      dictionary.push_back(input);
+    }
+
+    tree.insert_into(dictionary);
+    std::string searchStr;
+    std::cout<<"Enter the String that needs to be checked:"<<std::endl;
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:152]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+// Driver Function
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    std::cout<<"Enter the Strings in the Dictionary: (Press 1 to exit)"<<std::endl;
+    strlist dictionary;
+    dictionary={};
+    std::string input;
+    while(std::cin>>input){
+      if(input=="1") {break;}
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:146]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+// Driver Function
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    std::cout<<"Enter the Strings in the Dictionary: (Press 1 to exit)"<<std::endl;
+    strlist dictionary;
+    dictionary={};
+    std::string input;
+    while(std::cin>>input){
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:145]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    return v0.back();
+}
+
+// Driver Function
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    std::cout<<"Enter the Strings in the Dictionary: (Press 1 to exit)"<<std::endl;
+    strlist dictionary;
+    dictionary={};
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:143]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    return v0.back();
+}
+
+// Driver Function
+int main(void){
+    std::string Name = "Root";
+    BKTree tree(Name);
+    std::cout<<"Enter the Strings in the Dictionary: (Press 1 to exit)"<<std::endl;
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:141]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+            int iCost = v1[j] + 1;
+            int sCost = v0[j];
+            if (test[i] != comp[j])
+                sCost ++;
+
+            v1[j + 1] = min(dCost, iCost, sCost);
+        }
+        v0 = v1;
+    }
+
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:133]: Remove this line

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+        return b;
+    return c;
+}
+
+// Calculating the Levenshtein distance
+int BKTree::d(std::string& test, std::string& comp){
+    int m = test.length();
+    int n = comp.length();
+
+	std::vector< int > v0(n + 1);
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:114]: Inconsistency in indentation

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+        void search(Node* node, std::string& Name, int target_dist, strlist& suggestion_points);
+        int min(int& a, int& b, int& c);
+        int d(std::string& test, std::string& comp);
+        void delete_node(Node* node);
+};
+
+// Constructor to initialize the BK Tree
+BKTree::BKTree(std::string& Root_STR) {
+    root = new Node();
+    root->Name = Root_STR;
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:39]: Add space around the operator

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+void BKTree::insert(std::string& Name, Node*& node){
+
+    if (node == nullptr){
+        node = new Node();
+        node->Name = Name;
+    }
+    else {
+        int dist = d(Name, node->Name);
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:59]: Add space around the operator

**Comment by somya-kapoor** [comment]
@HarshCasper Are you working on it?

**Comment by HarshCasper** [comment]
Yes. I will resolve all issues by tomorrow.

**Comment by somya-kapoor** [comment]
@HarshCasper Are you working on it?

**Comment by HarshCasper** [comment]
I have updated the changes. Kindly review it @somya-kapoor

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+        return b;
+    return c;
+}
+
+// Calculating the Levenshtein distance
+int BKTree::d(std::string& test, std::string& comp){
+    int m = test.length();
+    int n = comp.length();
+	  std::vector< int > v0( n + 1 );
+	  std::vector< int > v1( n + 1 );
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:114]: no need for space after <, ( and before > ,)

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+int main(void){
+    std::string Name = "Root";
+    BKTree tree( Name );
+    std::cout << "Enter the Strings in the Dictionary: (Press 1 to exit)" << std::endl;
+    strlist dictionary;
+    dictionary = {};
+    std::string input;
+    while(std::cin >> input){
+      if(input == "1") { break; }
+      dictionary.push_back( input );
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:143]: no need for space after ( and before )

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    tree.insert_into( dictionary );
+    std::string searchStr;
+    std::cout << "Enter the String that needs to be checked:"<< std::endl;
+    std::cin >> searchStr;
+    strlist test = tree.search( searchStr, 2 );
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:150]: no need for space after ( and before )

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by HarshCasper** [comment]
I will add all the changes by Sunday.

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by HarshCasper** [comment]
I have added all the changes. Kindly take a look @somya-kapoor

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 26]
```diff
@@ -0,0 +1,161 @@
+/*
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:1]: File name should be Burkhard_Keller_Tree/Burkhard_Keller_Tree.cpp

File: Burkhard_Keller_Tree/burkhard_keller_tree.cpp [hunk 27]
```diff
@@ -0,0 +1,161 @@
+/* 
+Burkhard Keller Tree is a popular String Matching Algorithm which is used to
+perform spell checks based on the concept of Levenshtein distance.
+*/
+#include <iostream>
```
Comment (somya-kapoor) [inline Burkhard_Keller_Tree/burkhard_keller_tree.cpp:5]: leave a line before it

**Comment by somya-kapoor** [comment]
Please make the changes then it will ready to be merge

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by HarshCasper** [comment]
@somya-kapoor I will add all the changes by tomorrow.

**Comment by HarshCasper** [comment]
I have added the changes @somya-kapoor

**Comment by somya-kapoor** [comment]
please squash the commits

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by HarshCasper** [comment]
I will push the changes by Friday. Due to the Cyclone in my area, I was not able to work on it.

**Pull request closed without merging**